File: benchmark/bench_large.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <chrono>
#include "../generated/cpp_bench/cpp/bench_complex.hpp"

int main() {
    std::string fileName = "large_payload.bin";
    std::ifstream file(fileName, std::ios::binary | std::ios::ate);
    if (!file) {
        std::cerr << "Failed to open " << fileName << std::endl;
        return 1;
    }
    std::streamsize size = file.tellg();
    file.seekg(0, std::ios::beg);

    std::vector<uint8_t> buffer(size);
    if (!file.read((char*)buffer.data(), size)) {
        std::cerr << "Failed to read " << fileName << std::endl;
        return 1;
    }
    std::cout << "Read " << size << " bytes from " << fileName << std::endl;

    // Decode
    auto start = std::chrono::high_resolution_clock::now();
    WorldState world = WorldState::decode(buffer);
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsedDecode = end - start;
    std::cout << "Decode time: " << elapsedDecode.count() << "s" << std::endl;

    std::cout << "Decoded " << world.guilds.size() << " guilds" << std::endl;

    // Encode
    start = std::chrono::high_resolution_clock::now();
    std::vector<uint8_t> encodedData = world.encode();
    end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsedEncode = end - start;
    std::cout << "Encode time: " << elapsedEncode.count() << "s" << std::endl;
    std::cout << "Encoded size: " << encodedData.size() << " bytes" << std::endl;

    if (encodedData.size() != buffer.size()) {
        std::cout << "WARNING: Encoded size mismatch! Original: " << buffer.size() << ", New: " << encodedData.size() << std::endl;
    }

    return 0;
}

File: benchmark/rust/src/bench_large.rs
use std::fs::File;
use std::io::Read;
use std::time::Instant;

mod bench_complex_impl;
mod bench_complex_structs;
use bench_complex_impl::*;
use bench_complex_structs::WorldState;

fn main() {
    let file_name = "../../large_payload.bin"; // Path relative to where cargo run is executed (benchmark/rust)
    let mut file = match File::open(file_name) {
        Ok(f) => f,
        Err(_) => {
            // Try relative from project root if running from there
            match File::open("large_payload.bin") {
                Ok(f) => f,
                Err(e) => panic!("Failed to open {}: {}", file_name, e),
            }
        }
    };

    let mut data = Vec::new();
    file.read_to_end(&mut data).unwrap();
    println!("Read {} bytes from large_payload.bin", data.len());

    // Decode (1 run is usually enough for compiled languages like Rust for this size, but let's be consistent)
    // Actually, Rust is AOT compiled, so no JIT warmup like Java/JS needed.
    let start = Instant::now();
    let world = WorldState::decode(&data).expect("Failed to decode");
    // Access something to ensure no DCE (though decode typically parses fully)
    println!("Decoded {} guilds", world.guilds.len());
    let duration = start.elapsed().as_secs_f64();
    println!("Decode time: {:.4}s", duration);

    // Encode
    let start = Instant::now();
    let encoded_data = world.encode().expect("Failed to encode");
    let duration = start.elapsed().as_secs_f64();
    println!("Encode time: {:.4}s", duration);
    println!("Encoded size: {} bytes", encoded_data.len());

    if encoded_data.len() != data.len() {
        println!("WARNING: Encoded size mismatch! Original: {}, New: {}", data.len(), encoded_data.len());
    }
}
